bb_wrapper/wrapper/bb.py

```diff
@@ -101,8 +115,37 @@ def authenticate(self):
         }
         kwargs = dict(headers=header, verify=False, data=data)
 
-        response = requests.post(url, **kwargs)
-        response = self._process_response(response)
-        self.__access_token = response.data["access_token"]
-        self.__token_type = response.data["token_type"]
+        if self.data.access_token is None or force_auth is True:
+            response = requests.post(url, **kwargs)
+            response = self._process_response(response)
+            self.data.access_token = response.data["access_token"]
+            self.data.token_type = response.data["token_type"]
+            self.data.token_time = time.time()
+
+        return True
+
+    def __do_request(self, request_method, *args, **kwargs) -> requests.Response:
+        if self.data.access_token is None:
+            self.__authenticate()
+        else:
+            elapsed_time = time.time() - self.data.token_time
+            if elapsed_time >= self.TOKEN_TIME:
+                self.__authenticate(force_auth=True)
+
+        response = request_method(*args, **kwargs)
         return response
+
+    def _delete(self, url, headers=None) -> requests.Response:
+        return self.__do_request(super()._delete, url, headers)
+
+    def _get(self, url, headers=None) -> requests.Response:
+        return self.__do_request(super()._get, url, headers)
+
+    def _post(self, url, data, headers=None, use_json=True) -> requests.Response:
+        return self.__do_request(super()._post, url, data, headers, use_json)
+
+    def _put(self, url, data, headers=None, use_json=True) -> requests.Response:
+        return self.__do_request(super()._put, url, data, headers, use_json)
+
+    def _patch(self, url, data, headers=None, use_json=True) -> requests.Response:
+        return self.__do_request(super()._patch, url, data, headers, use_json)
```

Comment:
Com todas as refatorações feitas e códigos mais enxutos, não precisa mais também fazer o strategy pattern!

Já que teoricamente, seria necessário chamar o `__authenticate` antes de qualquer requisição


```suggestion
    def _delete(self, url, headers=None) -> requests.Response:
        self.__authenticate()
        return super()._delete(url, headers)

    def _get(self, url, headers=None) -> requests.Response:
        self.__authenticate()
        return super()._get(url, headers)

    def _post(self, url, data, headers=None, use_json=True) -> requests.Response:
        self.__authenticate()
        return super()._post(url, data, headers, use_json)

    def _put(self, url, data, headers=None, use_json=True) -> requests.Response:
        self.__authenticate()
        return super()._put(url, data, headers, use_json)

    def _patch(self, url, data, headers=None, use_json=True) -> requests.Response:
        self.__authenticate()
        return super()._patch(url, data, headers, use_json)
```